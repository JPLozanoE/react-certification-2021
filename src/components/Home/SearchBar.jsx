import React, { useContext, useState } from 'react';
import { useHistory } from 'react-router';
import { AppContext } from '../../state/AppContext';
import { types } from '../../types/types';
import { Bar, SearchButton, SearchInput } from './styles/SearchBar';

export const SearchBar = () => {
  const history = useHistory();
  const [input, setInput] = useState('');
  const { dispatch } = useContext(AppContext);

  const handleChange = (e) => {
    e.preventDefault();
    setInput(e.target.value);
  };

  const handleClick = () => {
    dispatch({
      type: types.setSelectedVideo,
      payload: null,
    });
    dispatch({
      type: types.changeSearchQuery,
      payload: input,
    });
    setInput('');
    history.push('/');
  };

  return (
    <Bar className="topnav" id="myTopnav">
      <SearchInput value={input} onChange={handleChange} placeholder="Buscar" />
      <SearchButton onClick={handleClick} name="buscar">
        buscar
      </SearchButton>
    </Bar>
  );
};
