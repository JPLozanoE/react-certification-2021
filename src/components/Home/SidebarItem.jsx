/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
import React, { useContext } from 'react';
import { useHistory } from 'react-router';
import { AppContext } from '../../state/AppContext';
import { types } from '../../types/types';
import { Button } from '../../__globalStyles';
import { VideoBox, DescriptionBox } from './styles/SidebarItem';

export const SidebarItem = ({ video, isFavorite = false, editFavorites = false }) => {
  const { dispatch } = useContext(AppContext);
  const history = useHistory();

  const handleClick = () => {
    dispatch({
      type: types.setSelectedVideo,
      payload: video,
    });
    if (editFavorites) {
      history.push(`/video/favorites/${video.id.videoId}`);
    } else {
      history.push(`/video/${video.id.videoId}`);
    }
  };

  const handleAddFavorite = (videoPayload) => {
    if (isFavorite) {
      dispatch({
        type: types.deleteFavoriteVideo,
        payload: videoPayload.id.videoId,
      });
      history.push('/favorites');
    } else {
      dispatch({
        type: types.addFavoriteVideo,
        payload: videoPayload,
      });
      dispatch({
        type: types.updateRecommendedVideos,
        payload: videoPayload.id.videoId,
      });
    }
  };

  const {
    snippet: { title, thumbnails },
  } = video;
  return (
    <>
      <VideoBox>
        <img
          onClick={handleClick}
          src={thumbnails.default.url}
          alt="img"
          aria-label={video.snippet.title}
        />
        <DescriptionBox>
          {title}
          <div>
            <Button onClick={() => handleAddFavorite(video)} type="button">
              {isFavorite ? 'Delete from favorites' : 'Add to Favorites'}
            </Button>
          </div>
        </DescriptionBox>
      </VideoBox>
      <hr />
    </>
  );
};
