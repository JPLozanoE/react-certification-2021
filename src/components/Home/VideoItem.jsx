/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
import React, { useContext } from 'react';
import { useHistory } from 'react-router';
import { AppContext } from '../../state/AppContext';
import { types } from '../../types/types';
import { VideoContainer, VideoTitle, VideoText } from './styles/VideoItem';
import { Button } from '../../__globalStyles';

export const VideoItem = ({ video, editFavorites, isFavorite }) => {
  const {
    dispatch,
    state: {
      auth: { isLogged },
    },
  } = useContext(AppContext);
  const history = useHistory();

  const handleVideoClick = (videoPayload) => {
    dispatch({
      type: types.setSelectedVideo,
      payload: videoPayload,
    });
    if (editFavorites) {
      history.push(`/video/favorites/${video.id.videoId}`);
    } else {
      history.push(`/video/${video.id.videoId}`);
    }
  };

  const handleAddFavorite = (videoPayload) => {
    if (editFavorites) {
      dispatch({
        type: types.deleteFavoriteVideo,
        payload: videoPayload.id.videoId,
      });
    } else {
      dispatch({
        type: types.addFavoriteVideo,
        payload: videoPayload,
      });
      dispatch({
        type: types.updateRecommendedVideos,
        payload: videoPayload.id.videoId,
      });
    }
  };

  return (
    <VideoContainer>
      <img
        width="100%"
        height="150px"
        style={{ objectFit: 'cover' }}
        src={video.snippet.thumbnails.high.url}
        alt="img"
        aria-label={video.snippet.title}
        onClick={() => handleVideoClick(video)}
      />
      <div style={{ padding: '10px' }}>
        <div>
          <VideoTitle onClick={() => handleVideoClick(video)}>
            {video.snippet.title}
          </VideoTitle>
          {isLogged && (
            <Button
              disabled={isFavorite && !editFavorites}
              onClick={() => handleAddFavorite(video)}
              type="button"
            >
              {editFavorites ? 'Delete from favorites' : 'Add to favorites'}
            </Button>
          )}
        </div>
        <VideoText>{video.snippet.channelTitle}</VideoText>
        <VideoText>Publicado el {video.snippet.publishTime}</VideoText>
      </div>
    </VideoContainer>
  );
};
