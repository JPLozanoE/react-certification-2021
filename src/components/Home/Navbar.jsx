import React, { useContext } from 'react';
import { SearchBar } from './SearchBar';
import { AppContext } from '../../state/AppContext';
import { types } from '../../types/types';
import {
  Bar,
  Links,
  CheckBoxWrapper,
  Text,
  CheckBoxLabel,
  CheckBox,
  DivDarkMode,
  DivSwitch,
} from './styles/NavBar';

export const Navbar = ({ setSelectedVideo }) => {
  const {
    state: {
      darkTheme,
      auth: { isLogged },
    },
    dispatch,
  } = useContext(AppContext);

  const handleClick = () => {
    dispatch({ type: types.setSelectedVideo, payload: null });
  };

  const handleLogout = () => {
    dispatch({ type: types.logout });
  };

  const handleToggleChange = () => {
    dispatch({
      type: types.toggleTheme,
    });
  };

  return (
    <Bar className="topnav" id="myTopnav">
      <Links onClick={handleClick} to="/">
        Inicio
      </Links>
      {/* <Links href="#" className="active" /> */}
      <SearchBar setSelectedVideo={setSelectedVideo} />
      <DivDarkMode>
        <Text href="#about">Oscuro</Text>
        <DivSwitch>
          <CheckBoxWrapper>
            <CheckBox
              defaultChecked={darkTheme}
              id="checkbox"
              type="checkbox"
              onChange={handleToggleChange}
            />
            <CheckBoxLabel htmlFor="checkbox" />
          </CheckBoxWrapper>
          {isLogged ? (
            <>
              <Links onClick={handleLogout} to="/">
                Logout
              </Links>
              <Links to="/favorites">Favorites</Links>
            </>
          ) : (
            <Links to="/auth">Login</Links>
          )}
        </DivSwitch>
      </DivDarkMode>
    </Bar>
  );
};
