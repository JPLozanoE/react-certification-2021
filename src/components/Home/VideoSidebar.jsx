import React from 'react';
import styled from 'styled-components';
import { SidebarItem } from './SidebarItem';

const SidebarLayout = styled.div`
  display: grid;
  grid-template-columns: 1fr;
`;

export const VideoSidebar = ({ videos, setSelectedVideo }) => {
  return (
    <SidebarLayout>
      {videos.map((video) => {
        if (video.id.videoId) {
          return (
            <SidebarItem
              setSelectedVideo={setSelectedVideo}
              key={video.etag}
              video={video}
            />
          );
        }
        return false;
      })}
    </SidebarLayout>
  );
};
