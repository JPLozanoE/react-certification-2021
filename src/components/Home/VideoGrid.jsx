import React from 'react';
import styled from 'styled-components';
import { VideoItem } from './VideoItem';

const VideoContainer = styled.div`
  display: grid;
  grid-template-columns: auto auto auto auto;

  @media only screen and (max-width: 1200px) {
    grid-template-columns: auto auto auto;
  }

  @media only screen and (max-width: 768px) {
    grid-template-columns: auto auto;
  }

  @media only screen and (max-width: 600px) {
    grid-template-columns: auto;
  }
`;

export const VideoGrid = ({ items, setSelectedVideo }) => {
  return (
    <VideoContainer>
      {items.map((video) => (
        <VideoItem setSelectedVideo={setSelectedVideo} {...video} key={video.etag} />
      ))}
    </VideoContainer>
  );
};
