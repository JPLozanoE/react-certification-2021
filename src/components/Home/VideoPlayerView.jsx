import React, { useContext } from 'react';
import {
  VideoContainer,
  ItemPlayerView,
  Videoiframe,
  Title,
  Description,
} from './styles/VideoPlayerView';
import { VideoSidebar } from './VideoSidebar';
import { AppContext } from '../../state/AppContext';
import { Button } from '../../__globalStyles';
import { types } from '../../types/types';

export const VideoPlayerView = ({ favorites }) => {
  // const [addedAsFavorite, setAddedAsFavorite] = useState(false);
  const {
    state: {
      videos,
      selectedVideo,
      favoriteVideos,
      auth: { isLogged },
    },
    dispatch,
  } = useContext(AppContext);

  const handleAddFavorite = (videoPayload) => {
    // setAddedAsFavorite((prevState) => !prevState);
    if (selectedVideo?.isFavorite) {
      dispatch({
        type: types.deleteFavoriteVideo,
        payload: videoPayload.id.videoId,
      });
      dispatch({
        type: types.deleteCurrentVideo,
        payload: videoPayload.id.videoId,
      });
    } else {
      dispatch({
        type: types.addFavoriteVideo,
        payload: videoPayload,
      });
      dispatch({
        type: types.updateRecommendedVideos,
        payload: videoPayload.id.videoId,
      });
      dispatch({
        type: types.addCurrentVideo,
      });
    }
  };

  const {
    id: { videoId },
    snippet: { title, description },
  } = selectedVideo;
  return (
    <VideoContainer>
      <ItemPlayerView>
        <Videoiframe
          width="100%"
          src={`https://www.youtube.com/embed/${videoId}`}
          frameBorder="0"
          allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
          allowFullScreen
          title="Embedded youtube"
        />
        {isLogged && (
          <Button onClick={() => handleAddFavorite(selectedVideo)} type="button">
            {selectedVideo?.isFavorite ? 'Delete from favorites' : 'Add to favorites'}
          </Button>
        )}
        <Title>{title}</Title>
        <Description>{description}</Description>
      </ItemPlayerView>
      <ItemPlayerView>
        <VideoSidebar
          videos={favorites ? favoriteVideos : videos}
          editFavorites={favorites}
        />
      </ItemPlayerView>
    </VideoContainer>
  );
};
