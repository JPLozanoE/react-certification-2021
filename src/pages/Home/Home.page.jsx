import React, { useContext } from 'react';
import { VideoGrid } from '../../components/Home/VideoGrid';
import { AppContext } from '../../state/AppContext';

function HomePage() {
  const {
    state: { videos },
  } = useContext(AppContext);

  return <VideoGrid edit={false} items={videos} />;
}

export default HomePage;
