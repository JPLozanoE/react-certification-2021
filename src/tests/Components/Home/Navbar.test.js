/* eslint-disable react/jsx-filename-extension */
import React from 'react';
import { render } from '@testing-library/react';
import { Navbar } from '../../../components/Home/Navbar';

describe('Navbar', () => {
  test('Should render properly', () => {
    const component = render(<Navbar />);
    expect(component).toMatchSnapshot();
  });
});
