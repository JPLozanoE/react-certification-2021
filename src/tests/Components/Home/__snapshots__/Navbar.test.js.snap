// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar not being logged in Should render properly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-hKFxyN hcZOaO topnav"
        id="myTopnav"
      >
        <a
          class="sc-eCApnc puTUN"
          href="/"
        >
          Home
        </a>
        <div
          class="sc-bdnxRM hqtFRX topnav"
          id="myTopnav"
        >
          <input
            class="sc-gtsrHT gJxAMF"
            placeholder="Enter your search..."
            value=""
          />
          <button
            class="sc-dlnjwi dVkyC"
            name="search"
          >
            Search
          </button>
        </div>
        <div
          class="sc-pNWdM jNockk"
        >
          <p
            class="sc-jSFjdj jQtSio"
            href="#about"
          >
            Dark mode
          </p>
          <div
            class="sc-jrsJWt cmODpM"
          >
            <div
              class="sc-gKAaRy bUbtZe"
            >
              <input
                checked=""
                class="sc-fujyAs drCHme"
                id="checkbox"
                type="checkbox"
              />
              <label
                class="sc-iCoGMd eIFASq"
                for="checkbox"
              />
            </div>
            <a
              class="sc-eCApnc puTUN"
              href="/auth"
            >
              Login
            </a>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-hKFxyN hcZOaO topnav"
      id="myTopnav"
    >
      <a
        class="sc-eCApnc puTUN"
        href="/"
      >
        Home
      </a>
      <div
        class="sc-bdnxRM hqtFRX topnav"
        id="myTopnav"
      >
        <input
          class="sc-gtsrHT gJxAMF"
          placeholder="Enter your search..."
          value=""
        />
        <button
          class="sc-dlnjwi dVkyC"
          name="search"
        >
          Search
        </button>
      </div>
      <div
        class="sc-pNWdM jNockk"
      >
        <p
          class="sc-jSFjdj jQtSio"
          href="#about"
        >
          Dark mode
        </p>
        <div
          class="sc-jrsJWt cmODpM"
        >
          <div
            class="sc-gKAaRy bUbtZe"
          >
            <input
              checked=""
              class="sc-fujyAs drCHme"
              id="checkbox"
              type="checkbox"
            />
            <label
              class="sc-iCoGMd eIFASq"
              for="checkbox"
            />
          </div>
          <a
            class="sc-eCApnc puTUN"
            href="/auth"
          >
            Login
          </a>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
